bus-mapping/src/circuit_input_builder/input_state_ref.rs

```diff
@@ -635,56 +624,22 @@ impl<'a> CircuitInputStateRef<'a> {
                 value_prev: sender_balance_prev,
             },
         )?;
-
-        let (_found, receiver_account) = self.sdb.get_account(&receiver);
-        let receiver_balance_prev = receiver_account.balance;
-        let receiver_balance = receiver_account.balance + value;
-        self.push_op_reversible(
-            step,
-            AccountOp {
-                address: receiver,
-                field: AccountField::Balance,
-                value: receiver_balance,
-                value_prev: receiver_balance_prev,
-            },
-        )?;
+        self.transfer_to(step, receiver, receiver_exists, opcode_is_create, value, true)?;
 
         Ok(())
     }
 
-    /// Same functionality with `transfer_with_fee` but with `fee` set zero.
-    pub fn transfer(
-        &mut self,
-        step: &mut ExecStep,
-        sender: Address,
-        receiver: Address,
-        receiver_exists: bool,
-        must_create: bool,
-        value: Word,
-    ) -> Result<(), Error> {
-        self.transfer_with_fee(
-            step,
-            sender,
-            receiver,
-            receiver_exists,
-            must_create,
-            value,
-            None,
-        )
-    }
-
     /// Transfer to an address. Create an account if it is not existed before.
     pub fn transfer_to(
         &mut self,
         step: &mut ExecStep,
         receiver: Address,
         receiver_exists: bool,
-        must_create: bool,
+        opcode_is_create: bool,
         value: Word,
         reversible: bool,
     ) -> Result<(), Error> {
-        // If receiver doesn't exist, create it
-        if (!receiver_exists && !value.is_zero()) || must_create {
+        if !receiver_exists && (!value.is_zero() || opcode_is_create) {
```

Comment:
This condition slips through the old PR that fixes all creation conditions.
https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1079/files/c0e6909d5135aa74e3ce415879d993a34050a823#r1336618956

Reply:
We have to think carefully about this condition. I would love input from reviewers.
- What is receiver_exists?
  - We can check if code_hash is 0. It is an artificial rule we added before.
  - Nonce is 0. Should we add this? address_collision in create has this check.
- opcode_is_create
  - originally called must_create. I think this covers 3 cases:
  - EOA transfer to create an account. Can we call this situtation "opcode_is_create"?
  - Create receiver with CREATE opcode
  - Create receiver with CREATE2 opcode

Reply:
[From geth impl.](https://github.com/ethereum/go-ethereum/blob/577be37e0e7a69564224e0a15e49d648ed461ac5/core/vm/evm.go#L192-L208), an account which doesn't exist will be created only when transfer value is non-zero.

But, no matter this account exsists or not, it'll be created again in CREATE/2 opcode. You can see [createAccount is called in create](https://github.com/ethereum/go-ethereum/blob/577be37e0e7a69564224e0a15e49d648ed461ac5/core/vm/evm.go#L448-L449) and then will find this

// createObject creates a new state object. If there is an existing account with
// the given address, it is overwritten and returned as the second return value.
in [createObject](https://github.com/ethereum/go-ethereum/blob/577be37e0e7a69564224e0a15e49d648ed461ac5/core/state/statedb.go#L626-L628)

The conclusion is the original condition is correct. In your fix, the account won't be re-created if the receiver exists.

Reply:
For the naming of opcode_is_create. I would vote is_create_account or force_create_account as you mentioned "EOA transfer to create an account", it doesn't fix the name.

Reply:
> an account which doesn't exist will be created only when transfer value is non-zero.

I think this is correct and it matches our behavior.

> no matter this account exists or not, it'll be created again in CREATE/2 opcode. 

This is incorrect. Three lines above your quoted [createAccount](https://github.com/ethereum/go-ethereum/blob/577be37e0e7a69564224e0a15e49d648ed461ac5/core/vm/evm.go#L448-L449) locates a nonzero nonce and non-empty hash check that early exits the function and returns an `ErrContractAddressCollision` error.

The [EIP-1014: Skinny CREATE2](https://eips.ethereum.org/EIPS/eip-1014) also states "if nonce or code (of the create2 destination address) is nonzero, then the create-operation fails."

So for this PR, the conclusion is that a CREATEX opcode does not guarantee an overwrite of an account.

----

But you might wonder, why the create opcode only checks non empty code hash and non zero nonce. What about storage root and balance? Can we have an account that has code="" and nonce=0, but storage root !=0 or balance !=0? The tl;dr is we can't, for non-trivial reason. 

I fell into a rabbit hole of a historical study and how the EIP was defined. Fortunately, Gary wrote a best summary of this issue. See https://github.com/ethereum/go-ethereum/pull/28666.

Reply:
Sorry for the confusion. Maybe we could put the link https://github.com/ethereum/go-ethereum/pull/28666 in our comment.

Reply:
I added to the comment in the constraint part, which might be the place people will check more frequently [77cfbcb](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1777/commits/77cfbcb9a3cdcf7c939996b339a2e7f48f9ffd30).

Reply:
> EOA transfer to create an account. Can we call this situtation "opcode_is_create"?

My opinion for `opcode_is_create` to represents `create an account` with tx.to = null is a bit confusing. How about just name it `is_create` to align with other place e.g. `call.is_create`?

Reply:
Sorry I missed the renaming part before. I fixed the renaming here [7449289](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1777/commits/74492894b0198af74cee971eb425805b0b5b14f4)